Add Broker Dispatcher to Server

ipsi opened the pull request:
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/broker/blob/master/.github/CONTRIBUTING.md) rules

#### What does this PR do?
When the server starts up, it should notify the Dispatcher that a new process is available to handle requests, and when it shuts down it should notify it that it is no longer available.

The Server will also notify the Dispatcher when a new connection is made, and when that connection is closed.

This also includes a TypeScript upgrade since some of our dependencies now depend on later versions of TypeScript.

#### Where should the reviewer start?
`dispatcher.js` contains the primary logic, with calls made from `server/index.js` and `server/socket.js`.

#### How should this be manually tested?


#### Any background context you want to provide?


#### Screenshots


#### Additional questions

ipsi committed:
Add Dispatcher calls to Server

When the server starts up, it should notify the Dispatcher that a new process is available to handle requests, and when it shuts down it should notify it that it is no longer available.

The Server will also notify the Dispatcher when a new connection is made, and when that connection is closed.

This also includes a TypeScript upgrade since some of our dependencies now depend on later versions of TypeScript.

ipsi committed:
TypeScript upgrade caused issues with a test

pavel-snyk commented on lib/dispatcher.js:
```diff
@@ -0,0 +1,187 @@
+  }
+}
+
+function hash(token) {
+  const shasum = crypto.createHash('sha256');
+  shasum.update(token);
+  return shasum.digest('hex');
+}
```
@ipsi, it makes sense to export this function, so I can re-use it as well on the client side.

pavel-snyk reviewed (dismissed):
LGTM.

Could we test dispatcher logic, so the flow is covered?

ipsi committed:
Send clientId to Dispatcher

Expect Client ID to be included in the Metadata sent to the Server, and forward it on to the Dispatcher.

ipsi committed:
Fix Content-Type and request body structure.

ipsi committed:
Set the server index correctly

snyksec commented:
:tada: This PR is included in version 4.138.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/snyk-broker/v/4.138.0)
- [GitHub release](https://github.com/snyk/broker/releases/tag/v4.138.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
